pip/resolve.py

```diff
@@ -173,6 +173,59 @@ def _check_skip_installed(self, req_to_install):
         self._set_req_to_reinstall(req_to_install)
         return None
 
+    def _get_abstract_dist_for(self, req):
+        assert self.require_hashes is not None, (
+            "require_hashes should have been set in Resolver.resolve()"
+        )
+
+        if req.editable:
+            return self.preparer.prepare_editable_requirement(
+                req, self.require_hashes
+            )
+
+        # satisfied_by is only evaluated by calling _check_skip_installed,
+        # so it must be None here.
+        assert req.satisfied_by is None
+        if not self.ignore_installed:
+            skip_reason = self._check_skip_installed(req)
+
+        if req.satisfied_by:
+            return self.preparer.prepare_installed_requirement(
+                req, self.require_hashes, skip_reason
+            )
+
+        upgrade_allowed = self._is_upgrade_allowed(req)
+        abstract_dist = self.preparer.prepare_linked_requirement(
+            req, self.session, self.finder, upgrade_allowed,
+            self.require_hashes
+        )
+
+        # NOTE
+        # The following portion is for determining if a certain package is
+        # going to be re-installed/upgraded or not and reporting to the user.
+        # This should probably get cleaned up in a future refactor.
+
+        # req.req is only available after unpack for URL pkgs repeat
+        # check_if_exists to uninstall-on-upgrade (#14)
```

Comment:
I copy-pasted it. I don't to change it as a part of this PR since I'm more concerned with moving stuff than improving things.